pkg/controller/utils/machine_helper_test.go

```diff
@@ -0,0 +1,280 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	awsproviderapi "sigs.k8s.io/cluster-api-provider-aws/pkg/apis/awsproviderconfig/v1beta1"
+)
+
+func TestUpdateAWSLBList(t *testing.T) {
+	clustername := "lbupdatetest"
+	sampleMachine := testutils.CreateMachineObj("master-0", clustername, "master", testutils.DefaultRegionName, testutils.DefaultAzName)
+	objs := []runtime.Object{&sampleMachine}
+	mocks := testutils.NewTestMock(t, objs)
+
+	oldLBList := []awsproviderapi.LoadBalancerReference{
+		{
+			// <clustername>-<id>-ext
+			Name: fmt.Sprintf("%s-%s-ext", clustername, "12345"),
+			Type: awsproviderapi.NetworkLoadBalancerType,
+		},
+		{
+			// <clustername>-<id>-int
+			Name: fmt.Sprintf("%s-%s-int", clustername, "12345"),
+			Type: awsproviderapi.NetworkLoadBalancerType,
+		},
+	}
+	newLBList := []awsproviderapi.LoadBalancerReference{
+		{
+			// Just something random!
+			Name: fmt.Sprintf("%s-%s-test", clustername, "12345"),
+			Type: awsproviderapi.NetworkLoadBalancerType,
+		},
+	}
+	// quickly make sure the test is going to measure an actual change
+	if len(oldLBList) != 2 || len(newLBList) != 1 {
+		t.Fatalf("Initial test conditions are unexpected. Old LB list should be 2 (got %d), New LB list should be 1 (got %d)", len(oldLBList), len(newLBList))
+	}
+	// decode spec
+	codec, err := awsproviderapi.NewCodec()
+	if err != nil {
+		t.Fatalf("Can't create decoder codec for AWS Provider API %v", err)
+	}
+	awsconfig := &awsproviderapi.AWSMachineProviderConfig{}
+	err = codec.DecodeProviderSpec(&sampleMachine.Spec.ProviderSpec, awsconfig)
+	if err != nil {
+		t.Fatalf("Can't decode sample Machine ProviderSpec %v", err)
+	}
+
+	err = updateAWSLBList(mocks.FakeKubeClient, oldLBList, newLBList, sampleMachine, awsconfig)
+	if err != nil {
+		t.Fatalf("Couldn't update AWS LoadBalancer List: %v", err)
+	}
+
+	machineInfo := types.NamespacedName{
+		Name:      sampleMachine.GetName(),
+		Namespace: sampleMachine.GetNamespace(),
+	}
+	// reload the object to make sure we're not just working with the "in-memory"
+	// representation, that being, the un-saved one.
+	err = mocks.FakeKubeClient.Get(context.TODO(), machineInfo, &sampleMachine)
+	if err != nil {
+		t.Fatalf("Couldn't reload the test machine: %v", err)
+	}
+
+	err = codec.DecodeProviderSpec(&sampleMachine.Spec.ProviderSpec, awsconfig)
+	if err != nil {
+		t.Fatalf("Can't decode sample Machine ProviderSpec %v", err)
+	}
+
+	if len(awsconfig.LoadBalancers) != 1 {
+		t.Fatalf("Expected to have only 1 LoadBalancer, but got %d", len(awsconfig.LoadBalancers))
+	}
+	if awsconfig.LoadBalancers[0].Name != newLBList[0].Name {
+		t.Fatalf("Expected LB name to be %s, but got %s", newLBList[0].Name, awsconfig.LoadBalancers[0].Name)
+	}
+
+	if awsconfig.LoadBalancers[0].Type != newLBList[0].Type {
+		t.Fatalf("Expected LB type to be %s, but got %s", string(newLBList[0].Type), string(awsconfig.LoadBalancers[0].Type))
+	}
+}
```

Comment:
we may also want a test here to ensure that no change is made if the two LB lists are the same

Reply:
> we may also want a test here to ensure that no change is made if the two LB lists are the same

Wouldn't this test still handle it? If `newLBList` is treated as the authoritative set of load balancers for the machines in question we would expect the code to always ensure that the machine load balancers are equal to that set. If it happens that `newLBList` and oldLBList` are the same, isn't the `updateAWSLBList` code effectively idempotent because at the end of the worst case, the machine load balancers will be removed and re-added with the same values?

Reply:
this makes good sense to me

Reply:
I may be misunderstanding, but it sounds like we're accepting the missing test based on knowing how the underlying code works. IOW: Yes, it's idempotent as written, but the tests shouldn't know that or rely on it. The point of these tests is to validate the visible behavior. If the code changes its algorithm later -- say, to do piecemeal add/remove rather than wholesale replace -- then an additional test like the one @csheremeta suggests are important to make sure that algorithm is correct.

Reply:
No, what I'm saying is that the behaviour of the method being tested is that the `newLBList` is always the authoritative source of what the Machine in question should have as its load balancer reality. Because of this, it is nonsensical to test what happens when old == new because at the end of the method, everything at all times will be set to the new list, and a test with a difference between old and new will cover just the same because of the end state this method promises: all will become equal to the new list's contents.

Reply:
We're agreeing that the method under test is always replacing with `newLBList`. If we're just interested in white-box/unit testing, what you're saying is legit.

But the point of writing tests we know are going to pass (they have to pass before the PR merges) is to catch bugs when the underlying code is changed. And to that end, there's value in writing the suggested test case, because it does cover a different _black-box_/interface behavior.

I'm not asking for a change here (or necessarily at all). I'm just asserting that it's not worthless.